src/main/java/org/terasology/ligthandshadow/barrier/MagicDomeSystem.java

```diff
@@ -28,119 +18,135 @@
+            float distance = df.length();
+//          logger.info("CharacerMoveInputEvent: position: {} - distance from O: {}, delta: {}", pos, distance, deltaDistance);
+            if (distance > WORLD_RADIUS && isAllowedInsideBarrier(player, domeEntity)) {
+//              logger.info("Sending player back inside!");
+                Vector3f impulse = df.normalize().negate();
+                impulse.mul(8);
+
+                player.send(new CharacterImpulseEvent(impulse));
+                player.send(new PlaySoundEvent(domeEntity.getComponent(MagicDome.class).hitSound, 2f));
+            }
+            if (distance < WORLD_RADIUS && !isAllowedInsideBarrier(player, domeEntity)) {
+//              logger.info("Sending player back outside");
+                Vector3f impulse = df.normalize();
+                impulse.mul(8);
+                player.send(new CharacterImpulseEvent(impulse));
 
-            if (deltaDistance > 0.2f) {
-//                logger.info("CharacerMoveInputEvent: position: {} - distance from O: {}, delta: {}", pos, distance, deltaDistance);
+                player.send(new PlaySoundEvent(domeEntity.getComponent(MagicDome.class).hitSound, 2f));
 
-                if (lastPos.length() < WORLD_RADIUS && lastPos.length() < pos.length() && distance > WORLD_RADIUS) {
-//                    logger.info("Sending player back inside!");
-                    Vector3f impulse = new Vector3f(pos).normalize().negate();
-                    impulse.mul(64.0f);
-                    impulse.y = 6;
+            }
+        }
+    }
 
-                    player.send(new CharacterImpulseEvent(impulse));
-                    player.send(new PlaySoundEvent(magicDomeEntity.getComponent(MagicDome.class).hitSound, 2f));
-                }
+    public EntityRef createBarrier(String prefab, Vector3ic location, String team) {
+        EntityRef barrier = entityManager.create(prefab, new Vector3f(location));
+        LocationComponent loc = barrier.getComponent(LocationComponent.class);
+        loc.setWorldScale(2 * WORLD_RADIUS * 1.01f);
+        MagicDome magicDome = barrier.getComponent(MagicDome.class);
+        magicDome.team = team;
+        barrier.addOrSaveComponent(loc);
+        barrier.addOrSaveComponent(magicDome);
+        return barrier;
+    }
 
-                if (lastPos.length() > WORLD_RADIUS && lastPos.length() > pos.length() && distance < WORLD_RADIUS) {
-//                    logger.info("Sending player back outside");
-                    Vector3f impulse = new Vector3f(pos).normalize();
-                    float verticalDiff = Math.sqrt(Math.abs(Math.sqrt(lastPos.y()) - Math.sqrt(domeVerticalTop.y())));
-                    float impulseY = (Math.abs(verticalDiff) / (float) WORLD_RADIUS) * 3f;
+    public boolean isAllowedInsideBarrier(EntityRef player, EntityRef barrier) {
+        return barrier.getComponent(MagicDome.class).team.equals(player.getComponent(LASTeamComponent.class).team);
```

Comment:
This is a nice one-liner, but also quite error prone to show a NPE some day. To simplify debugging, you may consider splitting this up in multiple lines:

```suggestion
        String barrierTeam = barrier.getComponent(MagicDome.class).team;
        String playerTeam = player.getComponent(LASTeamComponent.class).team;
        return barrierTeam.equals(playerTeam);
```